### DISCUSSION TITLE
Warm Starts using CBC

### DISCUSSION THREAD

**Discussion by U-Coding** [coin-or/pulp#411] [Q&A]
Hello,

I experimented with the wedding example https://github.com/coin-or/pulp/blob/master/examples/wedding_initial.py but somehow CBC seems to ignore the initial solution when I use "PULP_CBC_CMD(msg=True, warmStart=True)". 

However, when I use "PULP_CBC_CMD(msg=True, warmStart=True, **keepFiles=True**)" it seems to work ("Cbc0045I MIPStart provided solution with cost 12").

Is this a bug or some file-access problem on my system? I am using pulp 2.4 and cbc 2.10.3 (downloaded from https://ampl.com/products/solvers/open-source/#cbc).

I also tested with CPLEX 12.9.0: "CPLEX_CMD(msg=True, warmStart=True)" and it seems to work _without_ "keepFiles=True" ("MIP start 'm1' defined initial solution with objective 12.00").

Best regards

**Comment by pchtsp** [comment]
have you tried it with the CBC version that comes with pulp? with that version I'm able to get the warmStart read without issues. You should not need to provide `keepFiles=True`, as the example shows.

**Reply by U-Coding** [reply]
I tried with the CBC version that comes with pulp and I am getting the same result (only seems to be working with "keepFiles=True"). 

The difference I saw when looking at this again is that when keepFiles = True, a relative path is used for the mipstart file and when keepFiles = False, an absolute path is used, but it is starting with a dot. I think this prevents CBC from finding the mipstart file. (I am running the example via PyCharm Python Console.)

**This is what I get when keepFiles = False:**

Welcome to the CBC MILP Solver 
Version: 2.9.0 
Build Date: Feb 12 2015 
command line - C:\..\Anaconda3\lib\site-packages\pulp\apis\..\solverdir\cbc\win\64\cbc.exe C:\..\AppData\Local\Temp\1\9df4ff05b0424df9bf417732c2624063-pulp.mps mips C:\..\AppData\Local\Temp\1\9df4ff05b0424df9bf417732c2624063-pulp.mst branch printingOptions all solution C:\..\AppData\Local\Temp\1\9df4ff05b0424df9bf417732c2624063-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 23 COLUMNS
At line 24708 RHS
At line 24727 BOUNDS
At line 27941 ENDATA
Problem MODEL has 18 rows, 3213 columns and 15062 elements
Coin0008I MODEL read with 0 errors
**will open mipstart file .\C:\..\AppData\Local\Temp\1\9df4ff05b0424df9bf417732c2624063-pulp.mst.**
Continuous objective value is 12 - 0.00 seconds
Cgl0004I processed model has 18 rows, 3213 columns (3213 integer (3213 of which binary)) and 15062 elements
Cutoff increment increased from 1e-005 to 0.9999
Cbc0038I Initial state - 0 integers unsatisfied sum - 0
Cbc0038I Solution found of 12
Cbc0038I Before mini branch and bound, 3213 integers at bound fixed and 0 continuous
Cbc0038I Mini branch and bound did not improve solution (0.05 seconds)
Cbc0038I After 0.05 seconds - Feasibility pump exiting with objective of 12 - took 0.01 seconds
Cbc0012I Integer solution of 12 found by feasibility pump after 0 iterations and 0 nodes (0.05 seconds)
Cbc0001I Search completed - best objective 12, took 0 iterations and 0 nodes (0.05 seconds)
Cbc0035I Maximum depth 0, 0 variables fixed on reduced cost
Cuts at root node changed objective from 12 to 12
[..cut infos..]
Result - Optimal solution found
Objective value:                12.00000000
Enumerated nodes:               0
Total iterations:               0
Time (CPU seconds):             0.06
Time (Wallclock seconds):       0.06
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.11   (Wallclock seconds):       0.11
The choosen tables are out of a total of 3213:
('M', 'N')
('E', 'F', 'G')
('A', 'B', 'C', 'D')
('I', 'J', 'K', 'L')
('O', 'P', 'Q', 'R')

**This is what i get when keepFiles = True**

Welcome to the CBC MILP Solver 
Version: 2.9.0 
Build Date: Feb 12 2015 
command line - C:\..\Anaconda3\lib\site-packages\pulp\apis\..\solverdir\cbc\win\64\cbc.exe Wedding_Seating_Model-pulp.mps mips Wedding_Seating_Model-pulp.mst branch printingOptions all solution Wedding_Seating_Model-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 23 COLUMNS
At line 24708 RHS
At line 24727 BOUNDS
At line 27941 ENDATA
Problem MODEL has 18 rows, 3213 columns and 15062 elements
Coin0008I MODEL read with 0 errors
**will open mipstart file .\Wedding_Seating_Model-pulp.mst.**
mipstart values read for 3213 variables.
Continuous objective value is 12 - 0.02 seconds
Cgl0004I processed model has 18 rows, 3213 columns (3213 integer (3213 of which binary)) and 15062 elements
Cutoff increment increased from 1e-005 to 0.9999
**Cbc0045I mipstart provided solution with cost 12**
Cbc0006I The LP relaxation is infeasible or too expensive
Cuts at root node changed objective from 1.79769e+308 to -1.79769e+308
[..cut infos..]
Cgl0013I Postprocessed model is infeasible - possible tolerance issue - try without preprocessing
17 relaxed row infeasibilities - summing to 17
17 relaxed row infeasibilities - summing to 17
17 relaxed row infeasibilities - summing to 17
Result - Optimal solution found
Objective value:                12.00000000
Enumerated nodes:               0
Total iterations:               0
Time (CPU seconds):             0.07
Time (Wallclock seconds):       0.07
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.12   (Wallclock seconds):       0.12
The choosen tables are out of a total of 3213:

**Reply by tkralphs** [reply]
I guess this might be due to this bug, which is still outstanding: https://github.com/coin-or/Cbc/issues/32

**Reply by U-Coding** [reply]
Ok, thank you for your quick answers!

**Comment by hpswalters** [comment]
I ran into the same problem.  From my output, my best guess is that setting and fixing initial values requires writing these values to a file that is then read by the solver.  Therefore, keepfiles=True, forces the necessary file to be written.  I guess I'm keeping it
